Add miRDeep2

mschilli87 opened the pull request:
This PR depends on https://github.com/BIMSBbioinfo/guix-bimsb/pull/1.

This is my first (serious) attempt of packaging for Guix. So I'm sure there are several things to improve about this.

miRDeep2 doesn't have a `LICENSE` file on github but I double-checked with all authors that it is released under GLP3+. The github repository will reflect this soon.
However, as miRDeep2 depends on bowtie1 which is released under a non-free artistic license, I had to include it here.

The following points are the ones I'm the most uncertain about:

1. How can I make mirdeep2 use viennarna 2.2.10? If I `guix package -i viennarna` I get that one but mirdeep2 propagates 2.2.8.
2. Do I need native or propagated inputs anywhere (else)?
3. Does it matter where I placed the packages (is there any particular order)?
4. Did I break any style guidelines?

Let me know what to change and I'll amend my commit accordingly so we don't have any broken commit in the history.

---
Also here, I'd like to have @rekado have a look (in case that wasn't clear). :wink:

rekado commented:
Thanks for the patch!

For reproducibility I'd recommend using `git-fetch` to fetch the code from the tagged commit instead of relying on the on-demand tarballs generated by Github.  These tarballs are cached for a long time but will eventually be regenerated, causing the hash to change.  Using `git-fetch` with a tag bypasses this problem.

To get viennarna 2.2.10 we'd need to update the existing definition for viennarna.  It is currently bound to a package for version 2.2.8.  It's fine to replace the existing definition (I don't see a good reason to keep the older minor variant).

About style: you can avoid `(chdir "foo") ... (bar "../baz")` by using `(with-directory-excursion "foo" ...) ... (bar "baz")`.  Or you could use `copy-recursively` directly on "src".  Another nitpick is `(lambda () #t)` --- I'd replace it with the marginally shorter `(const #t)`.

Instead of propagating inputs I would try to either patch the source code to refer to the absolute path of the tools, or to wrap the executable(s) with `wrap-program`; you could wrap them in `PATH` (for other tools that are expected to be found on `PATH`) and `PERL5LIB` (for the Perl dependencies).  This way you avoid dumping dependent tools in the user's profile needlessly, which always increases the chance of conflicts.

mschilli87 commented:
Thx for the feedback. I'll have a look at the wrapping. I basically copy/pasted this together from existing packages, incl. the `(lambda () #t)`. :wink:

Regarding viennarna the issue is that there is already a definition for 2.2.10: `guix package -A viennarna` reports

```
viennarna       1.8.5   out     /home/mschilli/guix-bimsb-nonfree/bimsb/packages/bioinformatics-nonfree.scm:724:2
viennarna       2.2.8   out     /home/mschilli/guix-bimsb-nonfree/bimsb/packages/bioinformatics-nonfree.scm:674:2
viennarna       2.2.10  out     /home/mschilli/guix-bimsb-nonfree/bimsb/packages/bioinformatics-nonfree.scm:1097:2
```

and `guix package -i viennarna` installs 2.2.10. However, if I have this in my profile and then try to `guix package -i mirdeep2`, I get a conflict between the 2.2.10 one in my profile and the 2.2.8 propagated by miRDeep2. So I figured there might be an issue with some parts of Guix considerung 2.2.8 < 2.2.10 and some 2.2.10 < 2.2.8.
I guess by wrapping instead of propagating this conflict will be resolved but I still wonder why the 2.2.8 one is pulled in at all.

rekado commented:
Oh, I didn't see 2.2.10 there.  In that case it's easy: in Guix package definitions we specify dependencies by adding other package definitions to the inputs.  Instead of using the `(package ...)` form in the `inputs` field we refer to other packages through variable names.

Note the quasiquotation here:

```scheme
(inputs
 `(("whatever" ,foo)))
```

The backtick is syntax for `quasiquote`, the comma is short for `unquote`.  Unquote inserts the value of the expression that follows into the quoted expression.  Here `foo` is a variable that evaluates to a `package` value.

When you write `,viennarna` in a quasiquoted expression you're saying that the *value* bound to the variable `viennarna` should be pasted here.  Earlier we bound a package definition for ViennaRNA 2.2.8 to the variable name `viennarna`:

```scheme
(define-public viennarna
  (package
     (name "viennarna")
     (version "2.2.8")
     ...))
```

Further down on line 1097 (as reported by "guix package -A") you see that we define a different variable "viennarna-2.2.10", which is bound to a package expression for version 2.2.10.

To make this short: you need to refer to the variable `viennarna-2.2.10` instead of `viennarna`.

(Note that this is unusual.  In Guix we usually use the plain package name as the name of a variable that is bound to the latest version of that package.  I guess I didn't do this in this case out of laziness.)

mschilli87 commented:
> To make this short: you need to refer to the variable viennarna-2.2.10 instead of viennarna.

What I would have liked to have was: If there is no viennarna, get the latest, if there is already one, use it. But I guess this just prooves how I screwed up the hole idea of a functional package manager by propagating all those inputs. I'll change to the 2.2.10 and wrap it. Not sure if I'll find time tomorrow or this has to wait until next week but today I won't work on this anymore nor will I from Thursday to Sunday. :wink:

> (Note that this is unusual. In Guix we usually use the plain package name as the name of a variable that is bound to the latest version of that package. I guess I didn't do this in this case out of laziness.)

Shame on you! :stuck_out_tongue_winking_eye:

rekado commented:
> What I would have liked to have was: If there is no viennarna, get the latest, if there is already one, use it.

Yeah, that's not how this works :)

In Guix package definitions are Scheme values with direct links to other Scheme values; together they form a dense web of interdependent package values that are never ambiguous.  You'll always get exactly the value that is bound to a given variable name.

Note that for the command line interface the behaviour is just as you expect because what matters there is the *package name* (which is the same for all variants of viennarna); it'll automatically pick the latest version unless another version is specified.

mschilli87 commented:
I started working through your comments and except for the wrapping-part I think I'm done.
Please **don't merge** this yet as

1. I want to replace the propagated input by wrapping,
2. I'd prefer to squash all fixup commits into the first for the merge, and
3. I'd like to make sure everything works with latest Guix and rebased on latest `master`.

Personally I prefer 'byte-sized' commits for review, that's why I did not simply amend the original commit.

I started to work on the wrapping but ran into an error:

```
ERROR: In procedure %resolve-variable:
error: randfold: unbound variable
```
That one is quite easy to trace back as earlier I get

```
guix package: warning: failed to load '(bimsb packages staging)':
no code for module (gnu packages dlang)
```

I fetched the latest `master` of `guix-bimsb` and the only diff I got was

```diff
diff --git a/bimsb/packages/staging.scm b/bimsb/packages/staging.scm
index a289626..6623de4 100644
--- a/bimsb/packages/staging.scm
+++ b/bimsb/packages/staging.scm
@@ -207,7 +207,7 @@
        ("clang" ,clang-3.5)
        ("boost" ,boost)
        ("libuuid" ,util-linux)
-       ("pandoc" ,ghc-pandoc-1)
+       ("pandoc" ,ghc-pandoc)
        ("openssl" ,openssl)
        ("pam" ,linux-pam)
        ("zlib" ,zlib)))
```

which seems unrelated.
Could it be that a recent change on Guix removed that module leaving `guix-bimsb` broken?
Without `guix-bimsb` I don't have `randfold` (of course I use my `mirdeep2` branch) and therefor cannot test my `mirdeep2` package.

mschilli87 commented on bimsb/packages/bioinformatics-nonfree.scm:
```diff
@@ -1882,3 +1883,47 @@ MACE is a bioinformatics tool dedicated to analyze ChIP-exo data.")
+                                          "/bin")))
+               (with-directory-excursion "src"
+                 (copy-recursively "." target))
+               (with-output-to-file (string-append target "/../install_successful")
```
@rekado: Based on your comment in the first round of review

> you can avoid `(chdir "foo") ... (bar "../baz")` by using `(with-directory-excursion "foo" ...) ... (bar "baz")`.

I think you understood that the `../` referred to the `src` directory.
Note that this is not the case. That's also the reason this line is not part of the `(with-directory-excursion ...)` expression.

miRDeep2 comes with a `perl` script to download all dependencies, patching their Makefiles and installing them locally, and modifying `.bashrc` or `.zshrc` to set the `PATH` and `PERL5LIB`. In the end, it touches a file that is checked for *by the binaries at runtime* to ensure miRDeep2 was 'properly' installed.
As I wanted the dependencies and path modifications (or shebang-patching) to be handled by Guix, I did not use this script but followed the manual build instructions (adjusting on the way as needed). However, I need to make sure this file exists which is expected (hardcoded) one level *above* the `bin` directory containing the binaries (note that AFAIK, this directory doesn't need to be called `bin` but *must have a three character name*). That's why I append `/../` to `target` but if you prefer another way (i.e. some Scheme version of `dirname` or `(string-append (assoc-ref outputs "out") "/install_successful")`, let me know.
Also note that this will symlink an empty file as `~/.guix-profile/install_successful` which is weird but I guess won't ever introduce any actual conflict with another package. :wink:

I just wanted to clarify this to get your opinion because this is probably the weirdest version of `make check` I've ever come across. :laughing:

rekado replied:
If `make check` is essentially useless then we might as well disable it with `#:tests? #f`.  But maybe I misunderstood, because usually `make check` happens *before* `make install`, yet you create that file in the install phase.

Is any part of the code checking for that file to exist at runtime?  If so, I'd rather patch out this silly check.  What do you think?

mschilli87 replied:
Yeah you misunderstood, no `make check` there, just used it as an analogy. But doesn't matter.
I could patch it out but it's in several places across the code-base (why would it be factored out into an independent module? :wink:) and I was afraid that future updates will require pathichg other/additional places so maintaining this package would peobably be easier buy just doing what the software assumes should be done during installation.
If you still prefer patching, I can have a look how this is done in Guix and give it a shot.
But first, I'll figure out the wrapping part so we don't need propagated inputs.

rekado commented:
You should not get this error:

```
guix package: warning: failed to load '(bimsb packages staging)':
no code for module (gnu packages dlang)
```

because (gnu packages dlang) exists since Mon Sep 10 17:46:58 2018 +0200 (commit 98d6543f86d01486c2f6e808eedd97c601ba3e7a) in upstream Guix.  Are you using a recent version of Guix?  That module used to be named `(gnu packages ldc)`.

mschilli87 commented:
Could be a problem on my end, I had to move around some Guix stuff as we change some stuff on murphy.
I'll wait for this to be done, `guix pull && git fetch orgin && git rebase mirdeep2 origin/master` and take it from there.

mschilli87 commented:
I figured out my Guix issue (messed up PATH working around the missing filesystems) and replaced the input propagation by wrapping. I installed it and ran the tutorial but it fails with a bug fixed in miRDeep2's master branch. I'll make sure there is a new miRDeep2 release and will bump the version so I can fully test before we merge this.

I did come accross another file that needs to be installed one level above miRDeep.
If you are uncomfortable with the two files landing there, some pointers were to place them and how to
patch the source (there are quite some scripts) would be appreciated (though there is no better way of learning than going through the pain yourself). :wink:

@rekado: please give this another round of review so I can squash & rebase what we have so far.

---
**This is still not ready to merge!**

mschilli87 commented on bimsb/packages/bioinformatics-nonfree.scm:
```diff
@@ -1882,3 +1883,72 @@ MACE is a bioinformatics tool dedicated to analyze ChIP-exo data.")
+               (with-directory-excursion "src"
+                 (copy-recursively "." target))
+               (copy-file "Rfam_for_miRDeep.fa"
+                          (string-append  target "/../Rfam_for_miRDeep.fa"))
```
@rekado: This is the second file I was mentioning.

rekado replied:
Yeah, I don't like this.  I'd prefer to place these files in `$out/share/mirdeep2/` and patch the other files.  `share/<package>` is the canonical place to put "miscellaneous" package data.

mschilli87 replied:
Sure - I see why you don't like this. Any pointers as to what is the proper way to patch the sources?

mschilli87 commented on bimsb/packages/bioinformatics-nonfree.scm:
```diff
@@ -1882,3 +1883,72 @@ MACE is a bioinformatics tool dedicated to analyze ChIP-exo data.")
+    (inputs
+      `(("bowtie1" ,bowtie1)
+       ("perl-pdf-api2" ,perl-pdf-api2)
+       ("perl" ,perl)
```
Not sure if I need this having a perl-library as input.

rekado replied:
You would need "perl" among the inputs if any of the scripts have a Perl shebang (even if they use the /usr/bin/env trick). Guix would then automatically patch the shebang for you so that the particular variant of Perl from the inputs is used.

Since this package provides Perl scripts you will need Perl as an input.  The Perl module alone does not propagate Perl -- it just uses it to be built.

mschilli87 commented on bimsb/packages/bioinformatics-nonfree.scm:
```diff
@@ -1882,3 +1883,72 @@ MACE is a bioinformatics tool dedicated to analyze ChIP-exo data.")
+(define-public mirdeep2
+  (package
+    (name "mirdeep2")
+    (version "0.0.92")
```
This will need to be bumped before we squash, rebase & merge.

mschilli87 replied:
done in https://github.com/BIMSBbioinfo/guix-bimsb-nonfree/pull/1/commits/103168f2fe4de4096d22241effc70ef97b3be655

mschilli87 commented on bimsb/packages/bioinformatics-nonfree.scm:
```diff
@@ -1882,3 +1883,72 @@ MACE is a bioinformatics tool dedicated to analyze ChIP-exo data.")
+       (modify-phases %standard-phases
+         (delete 'configure)
+         (delete 'build)
+         (delete 'check)
```
Should I replace this by `test #f`?

rekado replied:
Yes.

mschilli87 replied:
done in https://github.com/BIMSBbioinfo/guix-bimsb-nonfree/pull/1/commits/37b0339af440904ca2388699ee583ba52f84f108

mschilli87 commented on bimsb/packages/bioinformatics-nonfree.scm:
```diff
@@ -1882,3 +1883,72 @@ MACE is a bioinformatics tool dedicated to analyze ChIP-exo data.")
+                              (,(getenv "PERL5LIB")))))
+                        '("make_html2.pl"
+                          "make_html.pl"
+                          "miRDeep2.pl"))
```
I only wrap files `use`-ing the dependency (according to `grep -l`), otherwise the list would be much longer.

rekado replied:
This is probably fine.

mschilli87 commented on bimsb/packages/bioinformatics-nonfree.scm:
```diff
@@ -1882,3 +1883,72 @@ MACE is a bioinformatics tool dedicated to analyze ChIP-exo data.")
+                          "miRDeep2_core_algorithm.pl"
+                          "miRDeep2.pl"
+                          "prepare_signature.pl"
+                          "quantifier.pl"))
```
Also here I used `grep -l` to determine which scripts need wrapping and which one don't. This could change with updates though (unlikely as miRDeep2 is pretty stable by now).

rekado replied:
You could also use `(find-files "." "\\.pl")` to get all Perl scripts.  This might be a little too much, but it wouldn't hurt.

mschilli87 commented on bimsb/packages/bioinformatics-nonfree.scm:
```diff
@@ -1882,3 +1883,72 @@ MACE is a bioinformatics tool dedicated to analyze ChIP-exo data.")
+       ("perl-pdf-api2" ,perl-pdf-api2)
+       ("perl" ,perl)
+       ("randfold" ,randfold)
+       ("viennarna" ,viennarna-2.2.10)))
```
I still would prefer to depend on `viennarna` as in 'latest' here to benefit from updates.
There are however several versions assigned to variables and used as inputs without comments regarding why they depend on a specific version. So I did not dare to touch this mess. :wink:

---
*edited to fix a typo that inverted the meaning* :grimacing:

rekado replied:
Yeah, I should rebind these variables so that `viennarna` is actually bound to the latest version.  Will do this after merging this patch.

mschilli87 replied:
@rekado: This is just a reminder to rebind `viennarna` after merging this PR. :wink:

mschilli87 reviewed (commented):
I commented on a few parts that I think deserve some extra attention but please comment on anything else you find.

mschilli87 commented on bimsb/packages/bioinformatics-nonfree.scm:
```diff
@@ -1882,3 +1883,95 @@ MACE is a bioinformatics tool dedicated to analyze ChIP-exo data.")
+                   "src/quantifier.pl"))
+               ;; patch script using ../Rfam_for_miRDeep.fa file
+               (substitute* "src/miRDeep2.pl"
+                 (("\\$\\{scripts\\}/../Rfam_for_miRDeep.fa")
```
Are these properly escaped?

rekado replied:
looks fine to me.  You could also escape the dots, because they really match any character once, but it's fine not to escape them.

mschilli87 replied:
I'll escape them just in case when finishing this PR.

mschilli87 replied:
done in https://github.com/BIMSBbioinfo/guix-bimsb-nonfree/pull/1/commits/2ccb7351e01e5bf94aab8bdf219ad160f3007dc6

mschilli87 commented on bimsb/packages/bioinformatics-nonfree.scm:
```diff
@@ -1882,3 +1883,95 @@ MACE is a bioinformatics tool dedicated to analyze ChIP-exo data.")
+                   (substitute* script
+                     (("\\$bn/../install_successful")
+                      (string-append out
+                                     "/share/mirdeep2/install_successful"))))
```
This will actually hard-code `/gnu/store/...-mirdeep2-.../share/mirdeep2/...` instead of `/var/guix/profiles/.../guix-profile/share/mirdeep2/...` - not sure if that's the right way to do it as clearly `~/guix-profile/share/mirdeep2/...` points to the right place.

rekado replied:
This is correct.  We only want references to /gnu/store, not references to profile links.

mschilli87 commented on bimsb/packages/bioinformatics-nonfree.scm:
```diff
@@ -1882,3 +1883,95 @@ MACE is a bioinformatics tool dedicated to analyze ChIP-exo data.")
+                          "quantifier.pl"))
+              #t))))))
+    (inputs
+      `(("bowtie1" ,bowtie1)
```
While I can use miRDeep2 without being able to run `bowtie` directly (thx to the wrapping), I still need a mapping index. So in most cases, I'll need to run `bowtie-build` (provided by `bowtie1` package) before I can run miRDeep2. So as a user I would be surprised that I have to `guix package -i bowtie1` even though I already did `guix package -i mirdeep2`.
Is that the way it should be (as strictly I *can* run miRDeep2 if I got an indexed genome elsewhere) or is this a case where the input should be propagated?

rekado replied:
I think it's better not to propagate.  It's better to let a user choose whether to install bowtie1 into the profile.  We'd propagate only when it's really necessary, e.g. when the package relies on PATH and cannot be patched, or when it is a language environment that requires modules to be installed alongside the package (e.g. for R, Python, Perl, or Guile packages).

mschilli87 reviewed (commented):
@rekado: I think this is ready for another round of review. I'm still awaiting feedback regarding the copyright of miRDeep2 before I can make a new release and finish this off, but I'd like to be ready here once I get an answer over there.

mschilli87 commented:
@rekado: I bumped the version and now I can run the test successfully further down than before.
However, I've hit an attempt to put some output files in `../<somedir>` relative to the install directory so I I'll have to check this out and patch the source accordingly.
Please let me know if my previous patching attempts need polishing as I'll likely do the same mistakes again now. :wink:

----
So ***still* not ready to merge**. :roll_eyes:

mschilli87 commented:
@rekado: One of miRDeeps scripts creates some files at runtime which are hardcoded to `"$(dirname $(which miRDeep2.sh))/indexes/Rfam_index*.ebwt"` which results in an attempt to write to `/gnu/store/...-profile/bin/indexes` which is a) obviously not permitted and b) completely crazy (why should indices go in a `bin` directory?). :boom: :hear_no_evil: :boom: 
I'll patch it to write to `~/.local/share/mirdeep2/indexes` instead (as that seems like the most common place to put such files were the user should have write permissions at runtime) unless you prefer another place.

mschilli87 commented on bimsb/packages/bioinformatics-nonfree.scm:
```diff
@@ -1882,3 +1883,100 @@ MACE is a bioinformatics tool dedicated to analyze ChIP-exo data.")
+               (substitute* "src/make_html.pl"
+                 (("`which miRDeep2.pl`") "`realpath ~/.local/share/mirdeep2`")
+                 (("mkdir \"\\$\\{scripts\\}indexes")
+                  "mkdir -p \"${scripts}indexes"))
```
Also here I wonder if I patch 'properly', especially w.r.t. escaping. :wink:

rekado replied:
this looks correct to me.

mschilli87 replied:
@rekado 
> Yes, sounds fine, though you'd need to make sure that this directory exists.

That's why I patch in the `-p` here to the corresponding `mkdir` call in miRDeep2 - so the directory will be created when needed as long as the user permissions allow the creation of `~/.local/share/mirdeep2`.
I figured that's safe to assume. :wink:

mschilli87 reviewed (commented):
This is now the first version that installs and runs the whole tutorial without error. :tada:
Once I get feedback regarding the open questions, I'd like to squash & rebase, test it on some real data, and then we can consider merging it. :wink:

rekado commented:
> @rekado: One of miRDeeps scripts creates some files at runtime which are hardcoded to `"$(dirname $(which miRDeep2.sh))/indexes/Rfam_index*.ebwt"` which results in an attempt to write to `/gnu/store/...-profile/bin/indexes` which is a) abviously not permitted and b) completely crazy (why should indices go in a `bin` directory?).   

Wow, that's strange.

> I'll patch it to write to `~/.local/share/mirdeep2/indexes` instead (as that seems like the most common place to put such files were the user should have write permissions at runtime) unless you prefer another place.

Yes, sounds fine, though you'd need to make sure that this directory exists.  If you were the developer, you'd probably not use ~/.local directly, but first check the relevant XDG_* env variables, as users may override these locations.

It would also be fine to write to the current working directory, but your way is prettier.

rekado commented:
Good job making this tool work!

If this is ready on your side I'll gladly merge it.

mschilli87 committed:
nonfree: Add mirdeep2.

mschilli87 commented:
@rekado: I fixed another small issue, rebased on latest master & squashed all commits into one.
So except for a real-word (*i.e.* non-tutorial data) run, this should finally be ready to merge now. :tada:

mschilli87 commented:
@rekado: I used this package for a project now without issues. I have to add that I only osed the `mapper.pl` and `quantifer.pl` modules with specific parameters so this is far from conclusive. But at the current state it is useful for me and I hope therefor also for others. So I'd suggest to merge this as it is now and if someone (incl. me) finds any issue, I'll take care of it then. :smiley:

mschilli87 commented:
@rekado: I just pushed a typo fix in my email address in the copyright comment. Now we should *really* be good to go. :wink:

rekado commented:
Thanks, merged!

I'll try to remember to rename viennarna later.

mschilli87 commented:
>  I'll try to remember to rename viennarna later.

I'll try to remind you in case you forget. :wink:
